src/backend/optimizer/plan/setrefs.c

```diff
@@ -1702,12 +1708,42 @@ set_foreignscan_references(PlannerInfo *root,
 
 	if (fscan->fdw_scan_tlist != NIL || fscan->scan.scanrelid == 0)
 	{
+		ListCell *cell;
+
 		/*
 		 * Adjust tlist, qual, fdw_exprs, fdw_recheck_quals to reference
 		 * foreign scan tuple
 		 */
 		indexed_tlist *itlist = build_tlist_index(fscan->fdw_scan_tlist);
 
+		foreach(cell, fscan->scan.plan.targetlist)
+		{
+			TargetEntry *tle;
+
+			tle = lfirst(cell);
+
+			if (IsA(tle->expr, Var))
+			{
+				Var *var;
+
+				var = (Var*) tle->expr;
+				if (var->varattno == GpForeignServerAttributeNumber)
+				{
+					FuncExpr	   *funcExpr;
+					RangeTblEntry  *rte;
+					Const 		   *relid;
+
+					rte = root->simple_rte_array[var->varno];
+					relid = makeConst(OIDOID, -1, InvalidOid, sizeof(Oid),
+									  ObjectIdGetDatum(rte->relid), false, true);
+					funcExpr = makeFuncExpr(6024, OIDOID, list_make1(relid), InvalidOid,
```

Comment:
6024 is what ?

Reply:
It is the oid of function gp_foreign_server_id, I will use a macro here